modules/graceful/server.go

```diff
@@ -31,6 +31,9 @@ var (
 // PerWriteWriteTimeout timeout for writes
 const PerWriteWriteTimeout = 5 * time.Second
```

Comment:
I think this does not sufficiently address the problem yet.

This code is not always executed a few times within the timeout frame, downloading with 50KB/s has for example 18 seconds between runs of the function on my end.

So I'd say there is need for a more elegant solution, or at least a higher default timeout.

Reply:
very exactly btw, confirmed this is the interval for a few minutes now ... thought it would be somewhat random at first.

Reply:
I mean this is more elegant as it takes account of previous writes through the previous deadline and the next write. 

The next step is accounting for how much is written directly and double counting for it. 

The other option is tuning those constants - I would guess that setting the first to 30s and maybe doubling the other would work.

Reply:
Yeah, my main issue is that the low values will never work if the function is not called before the deadline kicks in. It sounds cool if we could make sure it's called at least after a moment to calculate the deadline accordingly. On the other hand, I don't see a way to make sure it gets called if the connection stalls immediately (thinking of attacks or something) and the default timeout window is set very high.

I'll try to test if I can get some numbers about the situation, esp. connections that are x slow can still transfer y large files etc ...

Reply:
Okay, it looks like the situation improved drastically.
Downloading 100MB over 10KB/s fails with `2021-06-05 01:47:11 (10.0 KB/s) - Connection closed at byte 2621440. Giving up.` (Gitea timeout after `1m6.50294821s`)
but otherwise, this looks quote good, 100MB with 50KB/s worked fine, too.

Reply:
Ok. So if we up the baseline delay to 30s and adjust the formula to remove the halving we should be in business I guess.

Reply:
Ok I've adjusted things. 

Every write gets at least 30s. (pWWT)

Every kB written gets at least 2s. (pWWkBRate)

So either you get the previous deadline + the new perKbRate or 30s plus the new perKbRate.

Choose a minimum perKbTime that you will tolerate - the current is set probably too low. I would guess 10s per kB is the worst we should tolerate. 

I'll make these settings at the app.ini level.